### PR TITLE
wrap cudaStreamSynchronize calls

### PR DESCRIPTION
**Pull request by ngimel**
This is a first step towards creating context manager that errors out on synchronizing calls.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
Handling the ifdef in one place is a nice bonus. Will you add a lint to flag bare occurrences of these calls?

**Comment by ngimel** [comment]
Yes, will add a lint!

**Comment by byronyi** [comment]
@ngimel Any chance `cudaStreamSynchronize` could be completely removed from `at::nonzero`? We found it quite difficult to support for accelerators, [including NVIDIA GPU](https://github.com/mlcommons/training_results_v1.0/blob/master/NVIDIA/benchmarks/maskrcnn/implementations/pytorch/maskrcnn_benchmark/modeling/roi_heads/mask_head/mask_head.py#L35-L47).

**Comment by ezyang** [comment]
Given the existing semantics of the operation, removing the synchronization is not possible.

What we should do, however, is support JAX's extension to the nonzero API https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.nonzero.html where an explicit size can be specified, giving an upper bound to the number of nonzero entries that will be returned (and zero padded if there aren't enough). Can you file an issue for this?

**Comment by ngimel** [comment]
We would need this extension not only for nonzero, but also for indexing ops with mask, the most common situation when people encounter this particular sync is `out = x[mask]`

**Comment by byronyi** [comment]
> Given the existing semantics of the operation, removing the synchronization is not possible.
> 
> What we should do, however, is support JAX's extension to the nonzero API https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.nonzero.html where an explicit size can be specified, giving an upper bound to the number of nonzero entries that will be returned (and zero padded if there aren't enough). Can you file an issue for this?

Raised in #62320
